### PR TITLE
Clean up transparent block exchange if no response received

### LINKED ISSUE #429: Memory leak in MessageExchangeStore
It seems that memoryExchangeStore leaks.
Analyzing heap dump show that the leak is in  ExchangeByToken map and mainly concern observe request with block2.

The strange thing is most of exchanges have a response set, so it seems there are "complete" but not removed from the map... 

Here some old discussion about that :
https://dev.eclipse.org/mhonarc/lists/cf-dev/msg01404.html

#418 this PR aime to reproduce it via Junit.

### PR DESCRIPTION
**Pull request by sbernard31**
This aims to fix #429 (at least partially)
tests of #418 are also rebased on this branch. They could be used for more "checking". (#418 is not intended to be merged)

See #437 for more details about this choice of implementation.

### CODE REVIEW

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/BlockwiseClientSideTest.java [hunk 1]
```diff
@@ -84,9 +84,9 @@
 	private static final int MAX_RESOURCE_BODY_SIZE = 1024;
 	private static final int RESPONSE_TIMEOUT_IN_MS = 1000;
 	private static final int ERROR_TIMEOUT_IN_MS = 500;
-	// client retransmits after 200 ms
```
Comment (boaks) [inline californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/BlockwiseClientSideTest.java]: Any reason, why the comment is changed?

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java [hunk 2]
```diff
@@ -437,6 +451,40 @@ public void testQuickChangeAndTimeout() throws Exception {
...
+	public void testIncompleteBlock2Notification() throws Exception {
+		System.out.println("Observe with blockwise");
+		respPayload = generateRandomPayload(32);
+		byte[] tok = generateNextToken();
+
+		// Establish observe relation
+		respType = null; // first type is normal ACK
+		client.sendRequest(CON, GET, tok, ++mid).path(RESOURCE_PATH).observe(0).go();
+		client.expectResponse(ACK, CONTENT, tok, mid).storeObserve("A").storeETag("tag").payload(respPayload).go();
+		Assert.assertEquals("Resource has not added relation:", 1, testObsResource.getObserverCount());
```
Comment (boaks) [inline californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java]: I'm not sure, if we want to use a more uniform way, but if, I would have expected 

`assertThat("Resource has not added relation:", testObsResource.getObserverCount(), is(1)); `

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/Block1BlockwiseStatus.java [hunk 3]
```diff
@@ -32,6 +32,7 @@
 final class Block1BlockwiseStatus extends BlockwiseStatus {
 
 	private Request request;
+	private Exchange exchange;
```
Comment (boaks) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/Block1BlockwiseStatus.java]: Any reason, why the common stuff (`exchange` and `timeoutCurrentTranfer`) of `Block1BlockwiseStatus` and `Block2BlockwiseStatus` is not moved to the base class `BlockwiseStatus`?

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java [hunk 4]
```diff
@@ -493,7 +494,7 @@ public void testBlockwiseObserveNotInterruptedByOdlerNotificationWithoutBlockwis
 		server.sendResponse(CON, CONTENT).loadToken("OBS").observe(0).mid(++mid).payload(notifyPayload).go();
 		server.expectEmpty(ACK, mid).go();
 		// Check this one is discard.
-		Response response = request.waitForResponse(1000);
+		Response response = request.waitForResponse(150);
```
Comment (boaks) [inline californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: My experience with "small" timeouts and a "undeterministic" CPU allocation on a shared Buildserver are not that good.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/Block1BlockwiseStatus.java (hunk 3 above)
Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/Block1BlockwiseStatus.java]: I will move it in `BlockwiseStatus`.

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java (hunk 4 above)
Comment (sbernard31) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: I know what do you propose ?

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java (hunk 2 above)
Comment (sbernard31) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java]: All the tests in this class is written using :
Assert.assertEquals("Resource has not added relation:", 1, testObsResource.getObserverCount());

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/BlockwiseClientSideTest.java (hunk 1 above)
Comment (sbernard31) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/BlockwiseClientSideTest.java]: Just a mistake.

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java (hunk 2 above)
Comment (boaks) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveServerSideTest.java]: OK.

(Except of 2, which use the assertThat. Rise the question, which one should be used for future tests. But that's sure out of scope this PR.)

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java (hunk 4 above)
Comment (boaks) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: keep the 1000 (wasting some maven build time :-) ).

FMPOV, the check, if a response is accidentally received with a timeout, which seems to be too short for "shared build system", would not really check/ensure that.

**Comment by sbernard31** [comment]
I would like to use the same value for default `BLOCKWISE_STATUS_LIFETIME` in `BlockwiseLayer` and `NetworkConfigDefaults`. 5min, 30s or something else ?

**Comment by boaks** [comment]
I would prefer the 5min. 30s seems for me too short.

**Comment by sbernard31** [comment]
I remember that 5 minutes seemed too high for you, but ok for 5min.

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java (hunk 4 above)
Comment (sbernard31) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: That means increase the `blockwise_life_status` to more than 1s and so made all the test more than 1s longer this is really want you want ?

Comment (boaks) [reply californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: Yes.

Alternatively,  I (or you) creates a new branch and configure the unit test to execute that test repeated (using the RepeatingTestRunner) and set the "hudson_debug" job on it. So we will see, if that timing may cause an issue.

**Comment by sbernard31** [comment]
Is it better now ?

**Comment by boaks** [comment]
LGTM.

**Pull request merged**